Resolve #3

Wojtuz opened the pull request:
Resolve #3

Added a UI for Gyro visualisation/display of pitch and roll, float accessable via godot editor for UI testing.
Missing real time access to data from the rover (+ potencially icons for the visualisation)

wisniax commented:
# Good Job 🥇 

## But there are few things left:
-  Add class for the Deserializer based on [wiki](https://github.com/wisniax/RoverControlApp/wiki) to https://github.com/wisniax/RoverControlApp/blob/32f907a6119f2f826fba5ebf4ba4ebcac8d64702/Core/MqttClasses.cs#L12
( For the data you need/will need )
- Add respectable documentation entry for it with format, example JSON that you want to receive.
**NOTE:** It should coincide with ZED documentation linked in #3. 
In case it doesn't, finding a suitable conversion was also part of this task. (e.g. wrong type/unit)

## Minor details:
- If it's an UI change/improvement screenshot of that in PR (yes here, just ctrl+v the image) would be appreciated.
- Please, don't name all commits: _Add files via upload_ 🤣 
It is fine for the very first one. After that commit names should be appropriate to the changes ❤️

wisniax reviewed (changes requested):
Reason:
Point 1 & 2 in: https://github.com/wisniax/RoverControlApp/pull/10#issuecomment-1973748913

Wojtuz commented:
![image](https://github.com/wisniax/RoverControlApp/assets/83223696/e814c70d-c2bf-4f7c-ae0c-dd449f334669)

Wojtuz commented:
(potencially)
To be done:
-car like "check angles" yellow/red indicator when for example pitch will noticably impact movement speed/ability to move
-UI improvements
-slower update rate (less than every frame)

wisniax commented:
Hmm just slowing the rates down will suffice for now i think. Once that's done we can meet to see it working, and merge.
Keep up the good work!

Wojtuz commented:
Calculations and updates are now event based (not done every frame as they used to). I made a couple of UI improvements as well. This version has been tested using HiveMQ online client and I think it's ready to merge.
![image](https://github.com/wisniax/RoverControlApp/assets/83223696/0ad22350-d7a0-4dfa-a14a-778e468e2fad)

wisniax commented on project.godot:
```diff
@@ -199,5 +199,6 @@ app_backcapture_save={
 [rendering]
 
-renderer/rendering_method="gl_compatibility"
+renderer/rendering_method="mobile"
```
Why mobile? The main target is desktop

wisniax replied:
Why mobile? The main target is desktop

wisniax commented on MVVM/ViewModel/ZedMonitor.cs:
```diff
@@ -0,0 +1,117 @@
+    [Export]
+	Label rollDisplay;
+
+	double QuatX, QuatY, QuatZ, QuatW;
```
Is this really necessary here? Imo it should be passed as function argument below. Tip: U can use built-in Godot Quaterion.

AlvaroBajceps replied:
Another option: `Roll()` and `Pitch()` could be merged into `PullGyroscope()`, since they are both short and called right after.

wisniax reviewed (changes requested):
Please address the comments. After that we can merge ;)

Wojtuz commented:
![image](https://github.com/wisniax/RoverControlApp/assets/83223696/f2271ec1-fb5a-4ff2-9f5b-6d8d1de416d5)
UI changes after adding an error message (it would pop-up in case of an error regarding json deserialization, keep showing last known orientation and disappear when new (correct) info arrives).

Adressed your comments: 
Pitch(), Roll(), DisplayUpdate() and VisualisationUpdate() are no longer separate functions.
Changed rendering method back to compatibility. 
Modified quaternion calculations. I didn't manage to find a simple way to find roll and pitch with godot's built-in quaternion function.

wisniax reviewed (approved):
Lovely

AlvaroBajceps commented on MVVM/ViewModel/ZedMonitor.cs:
```diff
@@ -0,0 +1,91 @@
+    }
+    public void AngleUpdate(Quaternion Quat)
+    {
+        double rollDeg = (180.0 / Math.PI) * Math.Atan2(2 * (Quat.W * Quat.X + Quat.Y * Quat.Z), 1 - 2 * (Quat.X * Quat.X + Quat.Y * Quat.Y));
+        double pitchDeg = (180.0 / Math.PI) * Math.Asin(2 * (Quat.W * Quat.Y - Quat.Z * Quat.X));
```
## It's perfectly fine. ##
### Yet it can be simplified ###
`Godot.Quaterion` have `GetEuler` method. By retrieving angles in order ZYX you get [Nautical Angles](https://en.wikipedia.org/wiki/Euler_angles#Alternative_names) which represents roll, pitch and yaw (x,y,z axis in that order)

```suggestion
var eulerZYX = Quat.Normalized().GetEuler(EulerOrder.Zyx);
double rollDeg = Mathf.RadToDeg(eulerZYX.X);
double pitchDeg = Mathf.RadToDeg(eulerZYX.Y);
```

Wojtuz replied:
Resolved in [b84380b](https://github.com/wisniax/RoverControlApp/pull/10/commits/b84380bde7bbaec95036098780a888d8f6759cde). When I was trying to do it myself I mixed up the EulerOrder and assumed that I've done something else wrong. Thanks

AlvaroBajceps reviewed (changes requested):
Discussion regarding *MVVM/ViewModel/ZedMonitor.cs line 80* must be addressed before my approval.

Wojtuz commented on MVVM/View/ZedMonitor.tscn:
```diff
@@ -0,0 +1,106 @@
+scale = Vector2(0.297, 0.184)
+texture = ExtResource("1_esmsg")
+
+[node name="DVRollContainer" type="Label" parent="DataSection/DataVisualisation"]
```
Resolved in "[Changed containters housing visualization sprites from label to control](https://github.com/wisniax/RoverControlApp/pull/10/commits/8786525b703f335fbbe1bc91755a7f60d6d16230)"

Wojtuz commented on MVVM/ViewModel/ZedMonitor.cs:
```diff
@@ -0,0 +1,91 @@
+    }
+    private void DisplayUpdate(double rollDeg, double pitchDeg)
+	{
+        if(rollDeg == 0 && pitchDeg == 0)
```
Thought that both doubles being perfectly 0 is to improbable to take into account, but you are right. Fixed it in [950684f](https://github.com/wisniax/RoverControlApp/pull/10/commits/950684fee043a0b4565aeeb75a3627eee9c0262f). I also added a feature both showing time since last update to the user and showing information about data being outdated after a ceratain amount of time (setable in godot editor). [1072d25](https://github.com/wisniax/RoverControlApp/pull/10/commits/1072d25d49f58a91c64b48c5589f37cbd8cc9d30) Let me know if it's fine

AlvaroBajceps replied:
To be fair, it is improbable :)
### It's almost fine. ###
If `errorTime` can't be divided by `timerInterval` without remainder, error about gyro data being too old won't be logged (saved to .log file). 
> [MVVM/View/ZedMonitor.tscn](https://github.com/Wojtuz/RoverControlApp/blob/1431fceced6df88da1ca00c941195e6141e09b8c/MVVM/ViewModel/ZedMonitor.cs#L78-L92)

```cs
public void _on_timer_timeout()
{
    timeSLU += timerInterval;

    if (timeSLU == errorTime)
    {
        MainViewModel.EventLogger?.LogMessage($"ZedMonitor Error: gyro data >{errorTime} seconds old.");

    }
    if (timeSLU >= errorTime)
    {
        errorDisplay.Visible = true;
    }
    TimeUpdate(timeSLU);
}
```

Wojtuz commented:
UI changes:
![image](https://github.com/wisniax/RoverControlApp/assets/83223696/314e9cf5-eeff-43db-b94a-681559e86ae5)

Wojtuz commented on MVVM/ViewModel/ZedMonitor.cs:
```diff
@@ -55,18 +74,38 @@ public Task OnGyroscopeChanged(string subTopic, MqttApplicationMessage? msg)
+
+        if (timeSLU == errorTime)
+        {
+            GD.Print($"ZedMonitor Error: gyro data >{errorTime} seconds old.");
```
Fixed it in [03128b2](https://github.com/wisniax/RoverControlApp/pull/10/commits/03128b2bce88dd775536302c288b1540aac2f9b0). I don't even know why I wrote them like that considering that I used the EventLogger functions already. But it's fixed now.

AlvaroBajceps commented on MVVM/ViewModel/ZedMonitor.cs:
```diff
@@ -55,29 +74,50 @@ public Task OnGyroscopeChanged(string subTopic, MqttApplicationMessage? msg)
-            return new Quaternion();
+            MainViewModel.EventLogger?.LogMessage($"ZedMonitor Error (Something is wrong with json/deserialization): {e.Message}");
+            error = true;
+            return new Quaternion((float)Gyroscope.orientation.x, (float)Gyroscope.orientation.y, (float)Gyroscope.orientation.z, (float)Gyroscope.orientation.w);
```
I think `Gyroscope` will be null when json deserialization fails.
```suggestion
            return new Quaternion.Identity;
```

Wojtuz replied:
I checked it with a hivemq web client and gyroscope retains previous, correctly deserialized floats. However I agree that returning the same thing again is pointless especially now when we do nothing with that Quaternion. Changed in [c5fe7fa](https://github.com/wisniax/RoverControlApp/pull/10/commits/c5fe7fac5e284ba701c9156a7e23a590a999a66d)

AlvaroBajceps reviewed (changes requested):
Fix *MVVM/ViewModel/ZedMonitor.cs*
